[WTF] Add find32, find64, findFloat, and findDouble

Constellation opened the pull request:
#### e6956efe33b36966346311be4e5125e3ec2bb263
<pre>
[WTF] Add find32, find64, findFloat, and findDouble
<a href="https://bugs.webkit.org/show_bug.cgi?id=242843">https://bugs.webkit.org/show_bug.cgi?id=242843</a>

Reviewed by Sam Weinig and Darin Adler.

We rename memchr16 to find16 since it has slightly different semantics to memchr (usually, these function
takes bytes instead of string length, but find16 takes length, and this is better for WebKit).
And this patch adds find32, find64, findFloat, and findDouble, all of them are leveraging SIMD on ARM64.

                                            ToT                     Patched

biguint64-array-index-of-small         0.1911+-0.0042     ?      0.1911+-0.0046        ?
biguint64-array-index-of-large       884.3127+-0.3061     ^    637.4922+-4.8043        ^ definitely 1.3872x faster
bigint64-array-index-of-medium         0.4212+-0.0253     ^      0.3592+-0.0048        ^ definitely 1.1727x faster
float64-array-index-of-large         671.8128+-0.3408     ^    422.6900+-0.1360        ^ definitely 1.5894x faster
float32-array-index-of-medium          0.3688+-0.0053     ^      0.2792+-0.0045        ^ definitely 1.3211x faster
float64-array-index-of-small           0.1965+-0.0043            0.1946+-0.0049
int32-array-index-of-small             0.2035+-0.0149            0.1933+-0.0048          might be 1.0529x faster
int32-array-index-of-large           665.2647+-4.9674     ^    215.4366+-0.2330        ^ definitely 3.0880x faster
bigint64-array-index-of-large        884.8316+-1.1938     ^    638.3462+-0.1806        ^ definitely 1.3861x faster
bigint64-array-index-of-small          0.1954+-0.0046            0.1945+-0.0054
float64-array-index-of-medium          0.3687+-0.0053     ^      0.3166+-0.0052        ^ definitely 1.1645x faster
int32-array-index-of-medium            0.3776+-0.0204     ^      0.2783+-0.0056        ^ definitely 1.3568x faster
uint32-array-index-of-small            0.1938+-0.0049            0.1907+-0.0044          might be 1.0166x faster
uint32-array-index-of-large          663.0400+-0.2361     ^    215.2438+-0.0342        ^ definitely 3.0804x faster
uint32-array-index-of-medium           0.3657+-0.0051     ^      0.2749+-0.0047        ^ definitely 1.3303x faster
float32-array-index-of-large         666.3730+-0.6275     ^    215.2410+-0.0800        ^ definitely 3.0959x faster
biguint64-array-index-of-medium        0.4128+-0.0055     ^      0.3709+-0.0227        ^ definitely 1.1132x faster
float32-array-index-of-small           0.2012+-0.0060            0.1929+-0.0048          might be 1.0432x faster

* JSTests/microbenchmarks/bigint64-array-index-of-large.js: Added.
(test):
* JSTests/microbenchmarks/bigint64-array-index-of-medium.js: Added.
(test):
* JSTests/microbenchmarks/bigint64-array-index-of-small.js: Added.
(test):
* JSTests/microbenchmarks/biguint64-array-index-of-large.js: Added.
(test):
* JSTests/microbenchmarks/biguint64-array-index-of-medium.js: Added.
(test):
* JSTests/microbenchmarks/biguint64-array-index-of-small.js: Added.
(test):
* JSTests/microbenchmarks/int32-array-index-of-large.js: Added.
(test):
* JSTests/microbenchmarks/int32-array-index-of-medium.js: Added.
(test):
* JSTests/microbenchmarks/int32-array-index-of-small.js: Added.
(test):
* JSTests/microbenchmarks/uint32-array-index-of-large.js: Added.
(test):
* JSTests/microbenchmarks/uint32-array-index-of-medium.js: Added.
(test):
* JSTests/microbenchmarks/uint32-array-index-of-small.js: Added.
(test):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::indexOfImpl):
(JSC::genericTypedArrayViewProtoFuncIncludes):
(JSC::genericTypedArrayViewProtoFuncIndexOf):
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/text/StringCommon.cpp:
(WTF::memchr32AlignedImpl):
(WTF::memchr64AlignedImpl):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::memchr16):
(WTF::memchr32):
(WTF::memchr64):
(WTF::find):

Canonical link: <a href="https://commits.webkit.org/252945@main">https://commits.webkit.org/252945@main</a>
</pre>

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -437,27 +437,110 @@ ALWAYS_INLINE static size_t findInner(const SearchCharacterType* searchCharacter
+    constexpr size_t thresholdLength = 16;
+
+    size_t index = 0;
+    size_t runway = std::min<size_t>(thresholdLength, length);
```
No need for the `<size_t>` here since everything is `size_t`.

Constellation replied:
Fixed.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -437,27 +437,110 @@ ALWAYS_INLINE static size_t findInner(const SearchCharacterType* searchCharacter
 
-#if HAVE(MEMCHR16)
-WTF_EXPORT_PRIVATE const uint16_t* memchr16AlignedImpl(const uint16_t* pointer, uint16_t character, size_t length);
+ALWAYS_INLINE const uint8_t* find8(const uint8_t* pointer, uint8_t character, size_t length)
```
Since these functions all take pointers of distinct types, they could all just be a single overloaded `findScalar` instead of putting the scalar sizes into the function names. Not sure which we would prefer from a code readability and safety point of view; I am often an overloading fan but other people sometimes like the function names to be explicit. The overloading would be nice from a "usability in templates" point of view. Depending on our uses, I imagine even more overloading for other types of the same size, like `int32_t`, could be useful.

Constellation replied:
Currently, I prefer having `find8` etc. to make it explicit at callsite. But we can add `findScalar` layer later which calls these functions with `if-constexpr` switches :)

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -437,27 +437,110 @@ ALWAYS_INLINE static size_t findInner(const SearchCharacterType* searchCharacter
 
-ALWAYS_INLINE const uint16_t* memchr16(const uint16_t* pointer, uint16_t character, size_t length)
+    ASSERT(index < length);
+    return static_cast<const uint8_t*>(memchr(pointer + index, character, length - index));
```
I probably would have written a comment here saying something like, "we rely on `memchr` already having SIMD optimization, so we don’t have to write our own".

Constellation replied:
That's nice, done.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -437,27 +437,110 @@ ALWAYS_INLINE static size_t findInner(const SearchCharacterType* searchCharacter
     // For first check `threshold - (unaligned >> 1)` characters, we use normal loop.
     // This can (1) align pointer to 16byte size so that SIMD loop gets simpler and (2) handle cases
     // having a character in the beginning of the string efficiently.
     uintptr_t unaligned = (reinterpret_cast<uintptr_t>(pointer) & 0xf);
```
We don’t need those outer parentheses. I would omit them.

Constellation replied:
Done.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -437,27 +437,110 @@ ALWAYS_INLINE static size_t findInner(const SearchCharacterType* searchCharacter
+    constexpr size_t thresholdLength = 32;
+    static_assert(!(thresholdLength % (16 / sizeof(uint32_t))), "it should be 16byte aligned to make find32AlignedImpl simpler");
+
+    uintptr_t unaligned = (reinterpret_cast<uintptr_t>(pointer) & 0xf);
```
We don’t need those outer parentheses. I would omit them.

Constellation replied:
Done.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -437,27 +437,110 @@ ALWAYS_INLINE static size_t findInner(const SearchCharacterType* searchCharacter
+ALWAYS_INLINE const uint32_t* find32(const uint32_t* pointer, uint32_t character, size_t length)
+{
+    constexpr size_t thresholdLength = 32;
+    static_assert(!(thresholdLength % (16 / sizeof(uint32_t))), "it should be 16byte aligned to make find32AlignedImpl simpler");
```
I think "it should be" is not clear. I would say "length threshold should be". I would write "16-byte aligned" instead of "16byte aligned".

I would also like a comment explaining why the threshold should not be even higher. Did we measure it? It would be cool to leave a comment behind saying why we chose it so that someone later could choose an even better one, but has a head start understanding why we set it to 32 now.

Same thoughts on the similar code we already have in the 16-bit version above.

Constellation replied:
> I think "it should be" is not clear. I would say "length threshold should be". I would write "16-byte aligned" instead of "16byte aligned".

Fixed.

> I would also like a comment explaining why the threshold should not be even higher. Did we measure it? It would be cool to leave a comment behind saying why we chose it so that someone later could choose an even better one, but has a head start understanding why we set it to 32 now.

Yeah, this is picked from measurement to keep the smaller data case still the sane.

> Same thoughts on the similar code we already have in the 16-bit version above.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -437,27 +437,110 @@ ALWAYS_INLINE static size_t findInner(const SearchCharacterType* searchCharacter
+    if (runway == length)
+        return nullptr;
+
+    assert(index < length);
```
Oops: `assert` here instead of `ASSERT`. I assume that’s an accident.

Constellation replied:
Fixed.

darinadler commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -466,37 +549,107 @@ ALWAYS_INLINE const uint16_t* memchr16(const uint16_t* pointer, uint16_t charact
+#if CPU(ARM64)
+WTF_EXPORT_PRIVATE const float* findFloatAlignedImpl(const float* pointer, float target, size_t length);
+
+ALWAYS_INLINE const float* findFloat(const float* pointer, float target, size_t length)
```
I wonder whether clients in WebKit will more often want the real `findFloat` rather than the bitwise `==` version of `findFloat`. For hash tables, we found it was more useful to have bitwise comparison of floating point numbers for hash table keys rather than floating point equality. Maybe for some uses either would be OK, in which case we want the faster of the two. In other cases, it might be inconvenient that you can’t find a `NaN`.

Constellation replied:
This function is currently aligned to JS's indexOf behavior, and it requires floating point comparison, which includes `0 == -0` => `true`. And since we can just use `find32` if we would like to have bitwise-compared version, I think this `findFloat` is nice.

Constellation committed:
[WTF] Add find32, find64, findFloat, and findDouble
https://bugs.webkit.org/show_bug.cgi?id=242843

Reviewed by Sam Weinig and Darin Adler.

We rename memchr16 to find16 since it has slightly different semantics to memchr (usually, these function
takes bytes instead of string length, but find16 takes length, and this is better for WebKit).
And this patch adds find32, find64, findFloat, and findDouble, all of them are leveraging SIMD on ARM64.

                                            ToT                     Patched

biguint64-array-index-of-small         0.1911+-0.0042     ?      0.1911+-0.0046        ?
biguint64-array-index-of-large       884.3127+-0.3061     ^    637.4922+-4.8043        ^ definitely 1.3872x faster
bigint64-array-index-of-medium         0.4212+-0.0253     ^      0.3592+-0.0048        ^ definitely 1.1727x faster
float64-array-index-of-large         671.8128+-0.3408     ^    422.6900+-0.1360        ^ definitely 1.5894x faster
float32-array-index-of-medium          0.3688+-0.0053     ^      0.2792+-0.0045        ^ definitely 1.3211x faster
float64-array-index-of-small           0.1965+-0.0043            0.1946+-0.0049
int32-array-index-of-small             0.2035+-0.0149            0.1933+-0.0048          might be 1.0529x faster
int32-array-index-of-large           665.2647+-4.9674     ^    215.4366+-0.2330        ^ definitely 3.0880x faster
bigint64-array-index-of-large        884.8316+-1.1938     ^    638.3462+-0.1806        ^ definitely 1.3861x faster
bigint64-array-index-of-small          0.1954+-0.0046            0.1945+-0.0054
float64-array-index-of-medium          0.3687+-0.0053     ^      0.3166+-0.0052        ^ definitely 1.1645x faster
int32-array-index-of-medium            0.3776+-0.0204     ^      0.2783+-0.0056        ^ definitely 1.3568x faster
uint32-array-index-of-small            0.1938+-0.0049            0.1907+-0.0044          might be 1.0166x faster
uint32-array-index-of-large          663.0400+-0.2361     ^    215.2438+-0.0342        ^ definitely 3.0804x faster
uint32-array-index-of-medium           0.3657+-0.0051     ^      0.2749+-0.0047        ^ definitely 1.3303x faster
float32-array-index-of-large         666.3730+-0.6275     ^    215.2410+-0.0800        ^ definitely 3.0959x faster
biguint64-array-index-of-medium        0.4128+-0.0055     ^      0.3709+-0.0227        ^ definitely 1.1132x faster
float32-array-index-of-small           0.2012+-0.0060            0.1929+-0.0048          might be 1.0432x faster

* JSTests/microbenchmarks/bigint64-array-index-of-large.js: Added.
(test):
* JSTests/microbenchmarks/bigint64-array-index-of-medium.js: Added.
(test):
* JSTests/microbenchmarks/bigint64-array-index-of-small.js: Added.
(test):
* JSTests/microbenchmarks/biguint64-array-index-of-large.js: Added.
(test):
* JSTests/microbenchmarks/biguint64-array-index-of-medium.js: Added.
(test):
* JSTests/microbenchmarks/biguint64-array-index-of-small.js: Added.
(test):
* JSTests/microbenchmarks/int32-array-index-of-large.js: Added.
(test):
* JSTests/microbenchmarks/int32-array-index-of-medium.js: Added.
(test):
* JSTests/microbenchmarks/int32-array-index-of-small.js: Added.
(test):
* JSTests/microbenchmarks/uint32-array-index-of-large.js: Added.
(test):
* JSTests/microbenchmarks/uint32-array-index-of-medium.js: Added.
(test):
* JSTests/microbenchmarks/uint32-array-index-of-small.js: Added.
(test):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::indexOfImpl):
(JSC::genericTypedArrayViewProtoFuncIncludes):
(JSC::genericTypedArrayViewProtoFuncIndexOf):
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/text/StringCommon.cpp:
(WTF::memchr32AlignedImpl):
(WTF::memchr64AlignedImpl):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::memchr16):
(WTF::memchr32):
(WTF::memchr64):
(WTF::find):

Canonical link: https://commits.webkit.org/252945@main
